### PR TITLE
Improvements to PTB bucketed LSTM

### PR DESCRIPTION
**Pull request by mkolod**
- dropout support
  (needed for many stacked LSTM models, from language modeling to NMT,
   including the repro of the TensorFlow PTB language model)
- support of non-SGD/momentum optimizers
  (need to renove momentum option from optimizer param list)

### CODE REVIEW

File: example/rnn/cudnn_lstm_bucketing.py [hunk 1]
```diff
@@ -76,12 +78,15 @@ def train(args):
     if args.stack_rnn:
         stack = mx.rnn.SequentialRNNCell()
         for layer in range(args.num_layers):
+            dropout = 0.0
+            if layer < (args.num_layers - 1):
+                dropout = args.dropout
             stack.add(mx.rnn.FusedRNNCell(args.num_hidden, num_layers=1,
-                    mode='lstm', prefix='lstm_%d'%layer,
+                    mode='lstm', prefix='lstm_%d'%layer, dropout=dropout,
```
Comment (piiswrong) [inline example/rnn/cudnn_lstm_bucketing.py]: @mkolod Do you know if cudnn rnn drops the input or output or both? If it only drops output from non-final layer then this doesn't do anything.

Comment (mkolod) [inline example/rnn/cudnn_lstm_bucketing.py]: @piiswrong It's only applied in the middle of the stack. There's no dropout applied to the input of the first layer, or the output of the last layer. So, if manual explicit stacking of fused cells is used (as [here](https://github.com/dmlc/mxnet/blob/master/example/rnn/cudnn_lstm_bucketing.py#L80:L87)), rather than the whole stack being built as a black box (as [here](https://github.com/dmlc/mxnet/blob/master/example/rnn/cudnn_lstm_bucketing.py#L89:L90)), no CUDNN dropout will actually be applied. I guess for manually stacked FusedLSTM, this implies the need to manually apply dropout in MxNet, rather than counting on CUDNN to do it.

Comment (piiswrong) [inline example/rnn/cudnn_lstm_bucketing.py]: @szha Let's do the following:
change dropoutcell to normal cell instead of modifier. it's ok to break interface since it wasn't working anyway.

then modify unfuse to insert dropout in between

Comment (mkolod) [inline example/rnn/cudnn_lstm_bucketing.py]: @piiswrong you mean for the DropoutCell to inerit from BaseRNNCell rather than ModifierCell, correct?

**Comment by piiswrong** [comment]
Yes
